dev_tools/oneapi/add_host_function.sh

```diff
@@ -0,0 +1,305 @@
+                },
+            ::gko::syn::value_list<${list_list[bool]}>(),
+            ::gko::syn::value_list<${list_list[int]}>(),
+            ::gko::syn::value_list<${list_list[size_type]}>(),
+            ::gko::syn::type_list<${list_list[typename]}>(),
+            grid, block, dynamic_shared_memory, stream, ${var_input});
+    }
+    "
+}
+
+
+GLOBAL_KEYWORD="__global__"
+TEMPLATE_REGEX="^ *template <*"
+FUNCTION_START="^ *(template *<|${GLOBAL_KEYWORD}|void)"
+FUNCTION_NAME_END=".*\{.*"
+SCOPE_START="${FUNCTION_NAME_END}"
+SCOPE_END=".*\}.*"
+CHECK_GLOBAL_KEYWORD=".*${GLOBAL_KEYWORD}.*"
+FUNCTION_HANDLE=""
+DURING_FUNCNAME="false"
+ANAYSIS_FUNC=" *(template *<(.*)>)?.* (.*)\((.*)\)"
+START_BLOCK_REX="^( *\/\*| *\/\/)"
+END_BLOCK_REX="\*\/$| *\/\/"
+IN_BLOCK=0
+IN_FUNC=0
+STORE_LINE=""
+STORE_REGEX="__ *$"
+HOST_SUFFIX="_AUTOHOSTFUNC"
+EXTRACT_KERNEL="false"
+GINKGO_LICENSE_BEACON="******************************<GINKGO LICENSE>******************************"
+DURING_LICENSE="false"
+SKIP="false"
+MAP_FILE="map_list"
+rm "${MAP_FILE}"
+while IFS='' read -r line || [ -n "$line" ]; do
+    if [ "${EXTRACT_KERNEL}" = "false" ] && ([ "${line}" = "/*${GINKGO_LICENSE_BEACON}" ] ||  [ "${DURING_LICENSE}" = "true" ]); then
+        DURING_LICENSE="true"
+        if [ "${line}" = "${GINKGO_LICENSE_BEACON}*/" ]; then
+            DURING_LICENSE="false"
+            SKIP="true"
+        fi
+        continue
+    fi
+    # When do not need the license, do not need the space between license and other codes, neither.
+    if [ ${SKIP} = "true" ] && [ -z "${line}" ]; then
+        continue
+    fi
+    SKIP="false"
+    if [[ "$line" =~ ${STORE_REGEX} ]]; then
+        STORE_LINE="${STORE_LINE} ${line}"
+    elif [[ -n "${STORE_LINE}" ]]; then
+        echo "${STORE_LINE} ${line}"
+        STORE_LINE=""
+    else
+        echo "${line}"
+    fi
+    # echo "Handle___ ${line}"
+    if [[ "$line" =~ ${START_BLOCK_REX} ]] || [[ "${IN_BLOCK}" -gt 0 ]]; then
+        if [[ "$line" =~ ${START_BLOCK_REX} ]]; then
+            IN_BLOCK=$((IN_BLOCK+1))
+        fi
+        if [[ "$line" =~ ${END_BLOCK_REX} ]]; then
+            IN_BLOCK=$((IN_BLOCK-1))
+        fi
+        # echo ""
+        # echo "IN BLOCK ${IN_BLOCK}"
+        # output to new file
+        continue
+    fi
+    # echo "Handle ${line}"
+    # handle comments
+    if [[ "${line}" =~ $FUNCTION_START ]] || [[ $DURING_FUNCNAME = "true" ]]; then
+        # echo "line ${line}"
+        # echo "${FUNCTION_NAME_END}"
+        DURING_FUNCNAME="true"
+        FUNCTION_HANDLE="${FUNCTION_HANDLE} $line"
+        if [[ "${line}" =~ ${FUNCTION_NAME_END} ]]; then
+            # echo "end"
+            DURING_FUNCNAME="false"
+        fi
+        if [[ "${line}" =~ ${SCOPE_START} ]]; then
+            IN_FUNC=$((IN_FUNC+1))
+        fi
+        if [[ "${line}" =~ ${SCOPE_END} ]]; then
+            IN_FUNC=$((IN_FUNC-1))
+        fi
+        # output to new file
+        continue
+    fi
```

Comment:
In the end, all the previous things are skipped, i.e. not printed right? I
don't see any `echo`?

Reply:
it print the line in the upper else.
this is to skip those comments and not go into next block for handling add host